content/articles/building-spotify-recommendation-engine/index.md

```diff
@@ -0,0 +1,349 @@
+
+![equation](/engineering-education/building-spotify-recommendation-engine/equation.jpg)
+
+To understand it better, let us take an example of two items, item 1 and item 2, and two features $x_1$ and $x_2$, which define an item. The plot below represents item 1 and item 2 as vectors in a feature space.
+
+![cosine-similarity](/engineering-education/building-spotify-recommendation-engine/cosine-similarity.png)
+
+The lesser the angle between vectors more the cosine similarity.
+
+#### Collaborative Filtering
+
+Collaborative filtering systems use user-item interactions to generate recommendations. This means collaborative filtering uses similarities between users and items simultaneously to provide recommendations. To explain it better, let us refer to the diagram below.
+
+![collaborative-filtering](/engineering-education/building-spotify-recommendation-engine/collaborative-filtering.jpg)
+
+As we can see in the diagram, there are two users, user A, and user B. Both users have similar taste in music as both of them liked song-1 and song-2, but there is a song-3 which user A likes, but user B never listened to it. The system will recommend song-3 to user A based on these user-item interactions.
+
+As in this blog, the recommender system that we will implement is based on Content-Based Filtering. Therefore, we are going to limit our discussion of the Collaborative-Filtering system up to its definition.
+
+#### How to use Spotify Web API to fetch data
+
+For this blog, we will implement a custom playlist recommender for which we will require our current Spotify playlist data, using which we will further generate recommendations.
+For this purpose, Spotify provides a web API to create a custom application and fetch Spotify data. The Spotify Web API endpoints return JSON metadata about music artists, albums, and tracks directly from the Spotify Data Catalogue based on simple REST principles.
+
+![api](/engineering-education/building-spotify-recommendation-engine/web-api.jpg)
+
+**Simple steps to use Spotify Web API**
+
+Firstly open the Spotify Web API [dashboard](https://developer.spotify.com/dashboard/login). Log in using your Spotify account credentials. After logging in, your homepage will look something like this:
+
+![dashboard](/engineering-education/building-spotify-recommendation-engine/dashboard.png)
+
+Now click on create an application. This will generate a unique client id and password, which will be used further.
+And now open the application.
+
+![application-page](/engineering-education/building-spotify-recommendation-engine/spotify-api.png)
+
+Finally, go to the edit settings options and add your localhost URL. It will be used for user authentication later. And now we are all set to implement a Spotify-based playlist recommender system. 
+
+
+### Implementation
+
+To access the Spotify Web API, we will use a python-based library known as Spotipy, so let us first install this library. 
+
+```python
+!pip install spotipy
```

Comment:
Mention that you are working on a jupyter notebook. Else, make it generic and remove the !.